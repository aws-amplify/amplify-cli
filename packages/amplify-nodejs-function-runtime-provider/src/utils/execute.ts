import { existsSync } from 'fs-extra';
import { InvokeOptions } from './invokeOptions';
import path from 'path';
import _ from 'lodash';
import exit from 'exit';

//  copied from amplify-util-mock with slight modifications

// handler is a string like 'path/to/handler.func'
async function loadHandler(root: string, handler: string): Promise<Function> {
  const handlerParts = path.parse(handler);
  try {
    const handler = await import(path.join(root, handlerParts.dir, handlerParts.name));
    return handler[handlerParts.ext.replace('.', '')];
  } catch (err) {
    throw new Error(`Could not load lambda handler function due to ${err}`);
  }
}

export function invokeFunction(options: InvokeOptions) {
  return new Promise(async (resolve, reject) => {
    let returned = false;

    const context = {
      done(error: any, result: any) {
        if (!returned) {
          returned = true;
          if (error === null || typeof error === 'undefined') {
            context.succeed(result);
          } else {
            context.fail(error);
          }
        }
      },
      succeed(result: any) {
        returned = true;
        resolve(result);
      },
      fail(error: any) {
        returned = true;
        reject(error);
      },
      awsRequestId: 'LAMBDA_INVOKE',
      logStreamName: 'LAMBDA_INVOKE',
    };

    if (options.packageFolder) {
      const p = path.resolve(options.packageFolder);
      if (!existsSync(p)) {
        context.fail(`packageFolder ${options.packageFolder} does not exist`);
        return;
      }
      process.chdir(p);
    } else {
      context.fail('packageFolder is not defined');
      return;
    }

    if (!options.handler) {
      context.fail('handler is not defined');
      return;
    }

    if (options.context) {
      Object.assign(context, options.context);
    }

    const callback = (error: any, object: any) => {
      context.done(error, object);
    };

    const lambdaHandler = await loadHandler(options.packageFolder, options.handler);
    const { event } = options;
    try {
      const response = lambdaHandler(JSON.parse(event), context, callback);
      if (typeof response === 'object' && typeof response.then === 'function') {
        const result = await response;
        if (result !== undefined) {
          context.done(null, result);
        } else {
          context.done(null, null);
        }
      } else if (response !== undefined) {
        context.done(null, null);
      }
    } catch (e) {
      context.done(e, null);
    }
  });
}

process.on('message', async options => {
  try {
    const result = await invokeFunction(JSON.parse(options));
    process.stdout.write('\n');
    process.stdout.write(JSON.stringify({ result, error: null }));
  } catch (error) {
    process.send!(
      JSON.stringify({
        result: null,
        error: {
          type: 'Lambda:Unhandled',
          message: error.message,
        },
      }),
    );
    exit(1);
  }
  exit(0);
});
