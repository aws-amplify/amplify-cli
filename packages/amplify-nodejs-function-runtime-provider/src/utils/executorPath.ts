import * as path from 'path';

// This value is exported in it's own module so that it can be easily mocked in tests
export const executorPath = path.join(__dirname, 'execute.js');
