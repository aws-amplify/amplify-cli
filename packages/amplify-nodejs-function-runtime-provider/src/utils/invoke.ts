import { InvokeOptions } from './invokeOptions';
import path from 'path';
import * as execa from 'execa';

// copied from amplify-util-mock with slight modifications
export function invoke(options: InvokeOptions): Promise<any> {
  return new Promise((resolve, reject) => {
    try {
      let data: string = '';
      const lambdaFn = execa.node(path.join(__dirname, 'execute.js'), [], {
        env: options.environment || {},
      });
      lambdaFn.stdout.on('data', msg => {
        data += msg;
      });
      let inClosePromise: Promise<void> | null;
      const onClose = async () => {
        const lines = data.split('\n');
        if (lines.length > 1) {
          const logs = lines.slice(0, -1).join('\n');
          console.log(logs);
        }
        const lastLine = lines[lines.length - 1];
        try {
          const result = JSON.parse(lastLine);
          if (result.error) {
            reject(result.error);
          }
          resolve(result.result);
        } catch {
          resolve(lastLine);
        }
      };
      lambdaFn.on('close', () => {
        inClosePromise = onClose();
      });
      lambdaFn.catch(err => {
        const rejectWithClose = () => {
          if (inClosePromise) {
            inClosePromise.finally(() => {
              reject(err.message);
            });
          } else {
            reject(err.message);
          }
        };

        if (data.length > 0) {
          setTimeout(() => rejectWithClose(), 2000);
        } else {
          rejectWithClose();
        }
      });
      lambdaFn.send(JSON.stringify(options));
    } catch (e) {
      reject(e);
    }
  });
}
