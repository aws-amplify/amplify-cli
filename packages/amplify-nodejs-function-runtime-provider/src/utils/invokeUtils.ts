import path from 'path';
import * as execa from 'execa';

export type InvokeOptions = {
  packageFolder: string;
  handler: string;
  event: string;
  context?: object;
  environment?: { [key: string]: string };
};

export const getLambdaChildProcess = (environment: any, functionName: string = 'execute.js'): execa.ExecaChildProcess => {
  return execa.node(path.join(__dirname, functionName), [], {
    env: environment || {},
    extendEnv: false,
  });
};
