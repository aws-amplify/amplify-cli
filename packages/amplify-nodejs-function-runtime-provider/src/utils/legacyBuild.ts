import { getPackageManager } from 'amplify-cli-core';
import { BuildRequest, BuildResult, BuildType } from 'amplify-function-plugin-interface';
import execa from 'execa';
import * as fs from 'fs-extra';
import glob from 'glob';
import * as path from 'path';

// copied from the existing build-resources.js file in amplify-cli with changes for new interface
export async function buildResource(request: BuildRequest): Promise<BuildResult> {
  const resourceDir = request.service ? request.srcRoot : path.join(request.srcRoot, 'src');

  if (!request.lastBuildTimeStamp || isBuildStale(request.srcRoot, request.lastBuildTimeStamp, request.buildType, request.lastBuildType)) {
    installDependencies(resourceDir, request.buildType);
    if (request.legacyBuildHookParams) {
      runBuildScriptHook(request.legacyBuildHookParams.resourceName, request.legacyBuildHookParams.projectRoot);
    }
    return Promise.resolve({ rebuilt: true });
  }
  return Promise.resolve({ rebuilt: false });
}

function runBuildScriptHook(resourceName: string, projectRoot: string) {
  const scriptName = `amplify:${resourceName}`;
  if (scriptExists(projectRoot, scriptName)) {
    runPackageManager(projectRoot, undefined, scriptName);
  }
}

function scriptExists(projectRoot: string, scriptName: string) {
  const packageJsonPath = path.normalize(path.join(projectRoot, 'package.json'));
  if (fs.existsSync(packageJsonPath)) {
    const rootPackageJsonContents = require(packageJsonPath);

    return rootPackageJsonContents.scripts && rootPackageJsonContents.scripts[scriptName];
  }
  return false;
}

function installDependencies(resourceDir: string, buildType: BuildType) {
  runPackageManager(resourceDir, buildType);
}

function runPackageManager(cwd: string, buildType?: BuildType, scriptName?: string) {
  const packageManager = getPackageManager(cwd);

  if (packageManager === null) {
    // If no package manager was detected, it means that this functions or layer has no package.json, so no package operations
    // should be done.
    return;
  }

  const useYarn = packageManager.packageManager === 'yarn';
  const args = toPackageManagerArgs(useYarn, buildType, scriptName);
  try {
    execa.sync(packageManager.executable, args, {
      cwd,
      stdio: 'pipe',
      encoding: 'utf-8',
    });
  } catch (error) {
    if ((error as any).code === 'ENOENT') {
      throw new Error(`Packaging lambda function failed. Could not find ${packageManager} executable in the PATH.`);
    } else if (error.stdout?.includes('YN0050: The --production option is deprecated')) {
      throw new Error(`Packaging lambda function failed. Yarn 2 is not supported. Use Yarn 1.x and push again.`);
    } else {
      throw new Error(`Packaging lambda function failed with the error \n${error.message}`);
    }
  }
}

function toPackageManagerArgs(useYarn: boolean, buildType?: BuildType, scriptName?: string) {
  if (scriptName) {
    return useYarn ? [scriptName] : ['run-script', scriptName];
  }

  const args = useYarn ? [] : ['install'];

  if (buildType === BuildType.PROD) {
    args.push('--production');
  }

  return args;
}

function isBuildStale(resourceDir: string, lastBuildTimeStamp: Date, buildType: BuildType, lastBuildType?: BuildType) {
  const dirTime = new Date(fs.statSync(resourceDir).mtime);
  // If the last build type not matching we have to flag a stale build to force
  // a npm/yarn install. This way devDependencies will not be packaged when we
  // push to the cloud.
  if (dirTime > lastBuildTimeStamp || buildType !== lastBuildType) {
    return true;
  }
  const fileUpdatedAfterLastBuild = glob
    .sync(`${resourceDir}/**`)
    .filter(p => !p.includes('dist'))
    .filter(p => !p.includes('node_modules'))
    .find(file => new Date(fs.statSync(file).mtime) > lastBuildTimeStamp);
  return !!fileUpdatedAfterLastBuild;
}
