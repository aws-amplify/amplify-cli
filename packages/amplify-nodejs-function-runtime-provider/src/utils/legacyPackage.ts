import { $TSContext } from 'amplify-cli-core';
import { PackageRequest, PackageResult, ZipEntry } from 'amplify-function-plugin-interface';
import * as fs from 'fs-extra';
import glob from 'glob';
import * as path from 'path';

export async function packageResource(request: PackageRequest, context: $TSContext): Promise<PackageResult> {
  if (!request.lastPackageTimeStamp || request.lastBuildTimeStamp > request.lastPackageTimeStamp || request.currentHash) {
    const resourcePath = request.service ? path.join(request.srcRoot, '..') : path.join(request.srcRoot, 'src');
    const packageHash = !request.skipHashing ? await context.amplify.hashDir(resourcePath, ['node_modules']) : undefined;
    const zipEntries: ZipEntry[] = [];
    if (request.service) {
      const libGlob = glob.sync(resourcePath);
      const layerDirPath = path.join(request.srcRoot, '..', '..');
      const optPath = path.join(layerDirPath, 'opt');

      const conflicts: string[] = [];
      libGlob.forEach(lib => {
        const basename = path.basename(lib);
        if (fs.pathExistsSync(path.join(optPath, basename))) {
          conflicts.push(basename);
        }
      });

      if (conflicts.length > 0) {
        const libs = conflicts.map(lib => `"/${lib}"`).join(', ');
        const plural = conflicts.length > 1 ? 'ies' : 'y';
        context.print.warning(
          `${libs} subdirector${plural} found in both "/lib" and "/opt". These folders will be merged and the files in "/opt" will take precedence if a conflict exists.`,
        );
      }

      [...libGlob].forEach(folder => {
        if (fs.lstatSync(folder).isDirectory()) {
          zipEntries.push({
            packageFolder: folder,
          });
        }
      });
    } else {
      zipEntries.push({
        sourceFolder: resourcePath,
      });
    }
    return Promise.resolve({ packageHash, zipEntries });
  }
  return Promise.resolve({});
}
