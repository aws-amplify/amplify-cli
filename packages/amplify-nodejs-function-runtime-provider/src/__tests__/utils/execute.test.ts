import { invokeFunction } from '../../utils/execute';

describe('execute', () => {
  describe('async handler', () => {
    it('should succeed with a returned value', async () => {
      await expect(
        invokeFunction({
          packageFolder: __dirname,
          handler: 'handlers.asyncHandler',
          event: '{"succeed": true}',
          environment: {},
        }),
      ).resolves.toMatchSnapshot();
    });

    it('should succeed with a returned large payload', async () => {
      await expect(
        invokeFunction({
          packageFolder: __dirname,
          handler: 'handlers.asyncHandlerLargeData',
          event: '{"succeed": true}',
          environment: {},
        }),
      ).resolves.toMatchSnapshot();
    });

    it('should fail when throwing an error', async () => {
      await expect(
        invokeFunction({
          packageFolder: __dirname,
          handler: 'handlers.asyncHandler',
          event: '{"succeed": false}',
          environment: {},
        }),
      ).rejects.toMatchSnapshot();
    });
  });

  describe('callback handler', () => {
    it('should succeed with a returned value', async () => {
      await expect(
        invokeFunction({
          packageFolder: __dirname,
          handler: 'handlers.callbackHandler',
          event: '{"succeed": true}',
          environment: {},
        }),
      ).resolves.toMatchSnapshot();
    });

    it('should fail when returning an error', async () => {
      await expect(
        invokeFunction({
          packageFolder: __dirname,
          handler: 'handlers.callbackHandler',
          event: '{"succeed": false}',
          environment: {},
        }),
      ).rejects.toMatchSnapshot();
    });
  });

  describe('non-promise returned handler', () => {
    it('should resolve to null', async () => {
      await expect(
        invokeFunction({
          packageFolder: __dirname,
          handler: 'handlers.nonAsyncHandler',
          event: '{}',
          environment: {},
        }),
      ).resolves.toMatchSnapshot();
    });
  });
});
