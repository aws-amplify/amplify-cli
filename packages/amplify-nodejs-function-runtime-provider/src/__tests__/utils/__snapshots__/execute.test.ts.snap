// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`execute async handler should fail when throwing an error 1`] = `[Error: Fail]`;

exports[`execute async handler should succeed with a returned large payload 1`] = `
Array [
  Object {
    "__comment__": "Data generated from https://www.json-generator.com/",
    "_id": "5f1f207f0ebaf5a5238fe6fe",
    "about": "Nostrud mollit nulla minim sint eiusmod. Ex eu elit elit aliqua incididunt Lorem sit do qui dolor officia do elit laborum. Labore nulla adipisicing magna ipsum anim consequat culpa.
",
    "address": "357 Drew Street, Reinerton, Federated States Of Micronesia, 851",
    "age": 35,
    "balance": "$0",
    "company": "STEadsfasdELFAB",
    "email": "ayersladsfindsey@steelfab.com",
    "eyeColor": "brown",
    "favoriteFruit": "strawberry",
    "friends": Array [
      Object {
        "id": 0,
        "name": "Belinda Schmidt",
      },
      Object {
        "id": 1,
        "name": "Clare Oneill",
      },
      Object {
        "id": 2,
        "name": "Randolph Stewart",
      },
    ],
    "gender": "male",
    "greeting": "Hello, Ayers Lindsey! You have 5 unread messages.",
    "guid": "0db7fcdb-c6bb-4a27-ac45-0678fa79c61b",
    "index": 0,
    "isActive": true,
    "latitude": 23432.525794,
    "longitude": 234.121539,
    "name": "sdgfdasg Lindasfsadfdsey",
    "phone": "+1 (914) 515-2948",
    "registered": "2015-05-26T01:50:50 +07:00",
    "tags": Array [
      "mollit",
      "dolor",
      "dolore",
      "laboris",
      "Lorem",
      "ipsum",
      "tempor",
    ],
  },
  Object {
    "__comment__": "Data generated from https://www.json-generator.com/",
    "_id": "5f1f207f35cba573a8dd7872",
    "about": "Minim eu ex aute sit officia nulla aute veniam minim qui in ut. Cupidatat quis fugiat duis ex eu. Esse laborum dolor ea est tempor culpa aliqua minim mollit ad reprehenderit. Reprehenderit deserunt deserunt ipsum aute incididunt laboris officia tempor exercitation adipisicing sit excepteur cillum. Mollit est aute consectetur deserunt excepteur occaecat consectetur excepteur. Occaecat est consequat proident mollit occaecat in dolore aliquip dolore adipisicing exercitation enim quis. Quis ea Lorem aliqua amet laboris proident labore mollit laboris magna ipsum.
",
    "address": "463 Portland Avenue, Canterwood, Kentucky, 9776",
    "age": 37,
    "balance": "$1,179.06",
    "company": "LIMOZEN",
    "email": "savsdav@limozen.com",
    "eyeColor": "green",
    "favoriteFruit": "apple",
    "friends": Array [
      Object {
        "id": 0,
        "name": "Daniel Graves",
      },
      Object {
        "id": 1,
        "name": "Isabelle Juarez",
      },
      Object {
        "id": 2,
        "name": "Carson Travis",
      },
    ],
    "gender": "female",
    "greeting": "Hello, Marcy Powell! You have 2 unread messages.",
    "guid": "4d6b24c7-cb90-46a4-acf6-e2834ec877d3",
    "index": 1,
    "isActive": false,
    "latitude": 546.850388,
    "longitude": 6546450.078486,
    "name": "asdf cvfadsv",
    "phone": "+1 (932) 520-3033",
    "registered": "2015-02-10T11:11:58 +08:00",
    "tags": Array [
      "nisi",
      "anim",
      "reprehenderit",
      "est",
      "fugiat",
      "nisi",
      "reprehenderit",
    ],
  },
  Object {
    "__comment__": "Data generated from https://www.json-generator.com/",
    "_id": "5f1f207fd28d8f08d83fdd2d",
    "about": "Est esse duis ea consectetur est cillum exercitation in exercitation nulla incididunt enim aliquip officia. Pariatur nulla enim sunt minim consectetur. Mollit ut id nostrud minim ipsum adipisicing nostrud enim eu occaecat ad do. Exercitation excepteur amet non esse duis.
",
    "address": "697 Franklin Street, Manchester, Massachusetts, 1878",
    "age": 39,
    "balance": "$1,460.00",
    "company": "ZILLACON",
    "email": "alliemanning@zillacon.com",
    "eyeColor": "brown",
    "favoriteFruit": "apple",
    "friends": Array [
      Object {
        "id": 0,
        "name": "Richards Phillips",
      },
      Object {
        "id": 1,
        "name": "Roxie Larson",
      },
      Object {
        "id": 2,
        "name": "Lily Singleton",
      },
    ],
    "gender": "female",
    "greeting": "Hello, Allie Manning! You have 5 unread messages.",
    "guid": "5d2398c0-5059-48f2-b018-c200e9c5e4b8",
    "index": 2,
    "isActive": true,
    "latitude": 678.797917,
    "longitude": 156747.185075,
    "name": "Allie Manning",
    "phone": "+1 (995) 557-2312",
    "registered": "2017-06-07T08:32:52 +07:00",
    "tags": Array [
      "ad",
      "ex",
      "aliquip",
      "aliqua",
      "Lorem",
      "sunt",
      "Lorem",
    ],
  },
  Object {
    "__comment__": "Data generated from https://www.json-generator.com/",
    "_id": "5f1f207f2ad7eee108ac1e15",
    "about": "Minim cillum sint duis sint tempor duis do pariatur amet. Eu quis aliquip consequat Lorem elit esse quis ullamco sint. Cillum officia nostrud magna ea ea ullamco minim sit consequat. Duis cupidatat culpa aliquip proident irure voluptate. Irure voluptate exercitation fugiat anim voluptate tempor enim sit laborum laborum ex irure ipsum. In duis veniam est labore aliquip quis sint sunt.
",
    "address": "184 Dinsmore Place, Martell, Illinois, 6300",
    "age": 35,
    "balance": "$2,074.16",
    "company": "ZOINAGE",
    "email": "mathewsfischer@zoinage.com",
    "eyeColor": "green",
    "favoriteFruit": "banana",
    "friends": Array [
      Object {
        "id": 0,
        "name": "Browning Curtis",
      },
      Object {
        "id": 1,
        "name": "Johnson Sheppard",
      },
      Object {
        "id": 2,
        "name": "Leola Dunlap",
      },
    ],
    "gender": "male",
    "greeting": "Hello, Mathews Fischer! You have 3 unread messages.",
    "guid": "29416c92-c7ee-4318-8441-4a09aae07414",
    "index": 3,
    "isActive": true,
    "latitude": 58.689825,
    "longitude": -174.440921,
    "name": "Mathews Fischer",
    "phone": "+1 (901) 496-2861",
    "registered": "2014-12-14T09:56:21 +08:00",
    "tags": Array [
      "ex",
      "ipsum",
      "ullamco",
      "officia",
      "ullamco",
      "tempor",
      "nulla",
    ],
  },
  Object {
    "__comment__": "Data generated from https://www.json-generator.com/",
    "_id": "5f1f207f5cdc96631d5025b2",
    "about": "Elit laborum velit dolore enim Lorem reprehenderit est elit excepteur occaecat id. Do consequat reprehenderit quis consectetur occaecat do. Aute occaecat tempor adipisicing nulla minim cupidatat. Dolor sint labore aliquip nulla adipisicing officia sunt.
",
    "address": "795 Barlow Drive, Garberville, Alaska, 8267",
    "age": 30,
    "balance": "$1,824.35",
    "company": "PERKLE",
    "email": "crosbyacevedo@perkle.com",
    "eyeColor": "green",
    "favoriteFruit": "banana",
    "friends": Array [
      Object {
        "id": 0,
        "name": "Merritt Barr",
      },
      Object {
        "id": 1,
        "name": "Ortega Bradford",
      },
      Object {
        "id": 2,
        "name": "Lilia Griffith",
      },
    ],
    "gender": "male",
    "greeting": "Hello, Crosby Acevedo! You have 1 unread messages.",
    "guid": "3aca06e2-f71f-4eb1-b8d4-d05fb28acd62",
    "index": 4,
    "isActive": false,
    "latitude": 567.3401,
    "longitude": -85671.746076,
    "name": "Crosby Acevedo",
    "phone": "+1 (967) 482-2273",
    "registered": "2017-07-02T11:44:15 +07:00",
    "tags": Array [
      "consequat",
      "dolore",
      "voluptate",
      "Lorem",
      "nulla",
      "proident",
      "do",
    ],
  },
  Object {
    "__comment__": "Data generated from https://www.json-generator.com/",
    "_id": "5f1f207f2f05b6f7cf06f3fb",
    "about": "Aliquip laborum excepteur in commodo occaecat Lorem irure eiusmod fugiat eiusmod nulla. Cillum dolor id commodo irure commodo cillum non eiusmod aliquip non eiusmod dolore sunt. Fugiat occaecat aliqua cupidatat qui est.
",
    "address": "150 Melrose Street, Kempton, Northern Mariana Islands, 1583",
    "age": 22,
    "balance": "$1,010.04",
    "company": "TWIIST",
    "email": "guzmansellers@twiist.com",
    "eyeColor": "blue",
    "favoriteFruit": "apple",
    "friends": Array [
      Object {
        "id": 0,
        "name": "Ayala David",
      },
      Object {
        "id": 1,
        "name": "Shelley Dean",
      },
      Object {
        "id": 2,
        "name": "Shepard Ramos",
      },
    ],
    "gender": "male",
    "greeting": "Hello, Guzman Sellers! You have 2 unread messages.",
    "guid": "ec2760f4-b2e6-4ac1-b234-84fbcc892594",
    "index": 5,
    "isActive": true,
    "latitude": 5675.013715,
    "longitude": 9567650.472559,
    "name": "Guzman Sellers",
    "phone": "+1 (838) 536-2787",
    "registered": "2014-06-06T02:10:13 +07:00",
    "tags": Array [
      "nulla",
      "cupidatat",
      "nisi",
      "in",
      "consequat",
      "consequat",
      "est",
    ],
  },
  Object {
    "__comment__": "Data generated from https://www.json-generator.com/",
    "_id": "5f1f207fdf3d5bc79da6ba8f",
    "about": "Amet et culpa consequat reprehenderit esse eu ullamco. Do aute ullamco officia aute magna. Minim elit ut mollit cillum excepteur amet. Aliqua proident consequat id ex ipsum. Enim consectetur consequat magna ipsum. Magna sunt reprehenderit velit consectetur aute esse esse labore nulla. Sint eiusmod nulla nulla tempor enim excepteur nulla nisi tempor quis consequat duis qui dolore.
",
    "address": "118 Wortman Avenue, Irwin, Nebraska, 7501",
    "age": 37,
    "balance": "$3,970.91",
    "company": "EWAVES",
    "email": "stevensonmcmahon@ewaves.com",
    "eyeColor": "green",
    "favoriteFruit": "banana",
    "friends": Array [
      Object {
        "id": 0,
        "name": "Britney Jacobs",
      },
      Object {
        "id": 1,
        "name": "Earlene Faulkner",
      },
      Object {
        "id": 2,
        "name": "Kristine Dodson",
      },
    ],
    "gender": "male",
    "greeting": "Hello, Stevenson Mcmahon! You have 9 unread messages.",
    "guid": "6b6ea372-6186-4148-8d5f-94b2f657c36e",
    "index": 6,
    "isActive": false,
    "latitude": -567.379837,
    "longitude": 567.949371,
    "name": "Stevenson Mcmahon",
    "phone": "+1 (952) 463-2581",
    "registered": "2018-06-18T08:30:18 +07:00",
    "tags": Array [
      "fugiat",
      "nostrud",
      "irure",
      "culpa",
      "elit",
      "dolore",
      "dolor",
    ],
  },
  Object {
    "__comment__": "Data generated from https://www.json-generator.com/",
    "_id": "5f1f207fcb1eafad5a0a52db",
    "about": "Cillum nulla minim adipisicing fugiat occaecat ex exercitation exercitation magna est laborum deserunt. Laboris culpa nisi reprehenderit incididunt qui cupidatat aute eu adipisicing dolore est reprehenderit proident. Tempor dolore laboris duis incididunt magna est sint Lorem et deserunt ex.
",
    "address": "702 Albemarle Road, Clarktown, Maryland, 4564",
    "age": 38,
    "balance": "$1,337.40",
    "company": "QOT",
    "email": "dennismercado@qot.com",
    "eyeColor": "blue",
    "favoriteFruit": "strawberry",
    "friends": Array [
      Object {
        "id": 0,
        "name": "Sheila Tyler",
      },
      Object {
        "id": 1,
        "name": "Jocelyn Blair",
      },
      Object {
        "id": 2,
        "name": "Logan Battle",
      },
    ],
    "gender": "male",
    "greeting": "Hello, Dennis Mercado! You have 10 unread messages.",
    "guid": "1b71c0db-b46d-47a9-81b8-2bb903ca92c3",
    "index": 7,
    "isActive": false,
    "latitude": 56756.921085,
    "longitude": -56765.787235,
    "name": "Dennis Mercado",
    "phone": "+1 (805) 434-2179",
    "registered": "2015-12-15T06:08:50 +08:00",
    "tags": Array [
      "reprehenderit",
      "incididunt",
      "excepteur",
      "cillum",
      "cillum",
      "anim",
      "labore",
    ],
  },
  Object {
    "__comment__": "Data generated from https://www.json-generator.com/",
    "_id": "5f1f207f6c70a7f1c318243d",
    "about": "Nostrud magna cupidatat in eiusmod ad dolore consectetur sit et incididunt. Do aute adipisicing do reprehenderit commodo. Mollit consequat nulla ipsum Lorem commodo voluptate reprehenderit in eu. Esse tempor mollit enim esse enim commodo nostrud sit non incididunt occaecat reprehenderit quis. Incididunt cillum enim deserunt dolor magna proident culpa aliqua. Commodo pariatur irure tempor dolore dolor occaecat ea esse aliquip fugiat. Reprehenderit incididunt reprehenderit non consectetur ea pariatur velit minim esse minim magna id incididunt.
",
    "address": "686 Nolans Lane, Finzel, 56765756, 9162",
    "age": 39,
    "balance": "$1,254.68",
    "company": "COSMOSIS",
    "email": "patholcomb@cosmosis.com",
    "eyeColor": "blue",
    "favoriteFruit": "strawberry",
    "friends": Array [
      Object {
        "id": 0,
        "name": "Angela Mays",
      },
      Object {
        "id": 1,
        "name": "Bobbi Holland",
      },
      Object {
        "id": 2,
        "name": "Bonner Compton",
      },
    ],
    "gender": "female",
    "greeting": "Hello, Pat Holcomb! You have 4 unread messages.",
    "guid": "1ac0d1eb-7994-4a17-b15e-23cdc27ed9a4",
    "index": 8,
    "isActive": true,
    "latitude": -31.230469,
    "longitude": -84.802676,
    "name": "Pat Holcomb",
    "phone": "+1 (991) 572-3119",
    "registered": "2016-08-08T08:35:13 +07:00",
    "tags": Array [
      "qui",
      "incididunt",
      "ipsum",
      "minim",
      "exercitation",
      "veniam",
      "do",
    ],
  },
  Object {
    "__comment__": "Data generated from https://www.json-generator.com/",
    "_id": "5f1f207fc90c33e1d131ca20",
    "about": "Ut proident ex anim quis ipsum eu. Nisi amet consectetur officia cillum ut irure est aliquip elit enim. Reprehenderit dolore velit commodo ad quis ad. Dolore ullamco adipisicing esse aute dolore pariatur velit do deserunt mollit eiusmod aliqua eiusmod amet. Consequat in aliqua veniam aute exercitation laboris est aliqua nisi.
",
    "address": "958 Madeline Court, Dargan, South Dakota, 2477",
    "age": 29,
    "balance": "$2,735.19",
    "company": "COMVEY",
    "email": "bethdonovan@comvey.com",
    "eyeColor": "blue",
    "favoriteFruit": "apple",
    "friends": Array [
      Object {
        "id": 0,
        "name": "Jackie Mcfarland",
      },
      Object {
        "id": 1,
        "name": "Ryan Burgess",
      },
      Object {
        "id": 2,
        "name": "Sonia Salazar",
      },
    ],
    "gender": "female",
    "greeting": "Hello, Beth Donovan! You have 9 unread messages.",
    "guid": "c2a9bfc8-449a-4d33-829a-ceea3e0ca297",
    "index": 9,
    "isActive": true,
    "latitude": 40.959102,
    "longitude": -144.618281,
    "name": "Beth Donovan",
    "phone": "+1 (849) 402-2342",
    "registered": "2016-12-29T06:05:47 +08:00",
    "tags": Array [
      "velit",
      "dolor",
      "est",
      "mollit",
      "qui",
      "Lorem",
      "voluptate",
    ],
  },
  Object {
    "__comment__": "Data generated from https://www.json-generator.com/",
    "_id": "5f1f207f0235265f1b88c4cd",
    "about": "Proident consectetur incididunt pariatur veniam ipsum velit ut minim. Anim cupidatat veniam reprehenderit adipisicing. Sit adipisicing proident non aliqua aute duis veniam elit nostrud consectetur cupidatat. Reprehenderit anim Lorem do ullamco laboris incididunt cillum duis laboris ex nostrud do deserunt. Duis ullamco dolor aliquip cillum labore reprehenderit non magna nostrud. Dolor deserunt nostrud est elit aliquip nisi incididunt nisi ullamco consectetur nisi cupidatat laboris ex. Nisi non quis magna ullamco exercitation elit sit.
",
    "address": "500 Pine Street, Vienna, Ohio, 3620",
    "age": 38,
    "balance": "$3,589.77",
    "company": "MARTGO",
    "email": "melvacochran@martgo.com",
    "eyeColor": "brown",
    "favoriteFruit": "apple",
    "friends": Array [
      Object {
        "id": 0,
        "name": "Meyers Ward",
      },
      Object {
        "id": 1,
        "name": "Gibbs Reyes",
      },
      Object {
        "id": 2,
        "name": "Walsh Kidd",
      },
    ],
    "gender": "female",
    "greeting": "Hello, Melva Cochran! You have 2 unread messages.",
    "guid": "ed51c092-5045-4213-b591-b07168367697",
    "index": 10,
    "isActive": false,
    "latitude": -23.9445,
    "longitude": 86.277144,
    "name": "Melva Cochran",
    "phone": "+1 (800) 533-3599",
    "registered": "2019-03-24T08:22:03 +07:00",
    "tags": Array [
      "esse",
      "aute",
      "sint",
      "magna",
      "irure",
      "elit",
      "ad",
    ],
  },
  Object {
    "__comment__": "Data generated from https://www.json-generator.com/",
    "_id": "5f1f207f5f7c15d6db2bf8ff",
    "about": "Occaecat dolor officia veniam sit laboris et et proident voluptate adipisicing qui officia qui. Elit enim deserunt amet consequat qui ad sit. Sunt veniam cupidatat proident nulla. Magna deserunt quis aliqua enim excepteur ullamco in sunt do occaecat. Eu proident magna exercitation cillum exercitation anim anim laboris qui elit incididunt. Ad ex esse dolor commodo sunt sunt ex consectetur officia pariatur.
",
    "address": "129 Billings Place, Lookingglass, Vermont, 9799",
    "age": 30,
    "balance": "$1,580.54",
    "company": "BICOL",
    "email": "robinrodriguez@bicol.com",
    "eyeColor": "green",
    "favoriteFruit": "strawberry",
    "friends": Array [
      Object {
        "id": 0,
        "name": "Tate Ortiz",
      },
      Object {
        "id": 1,
        "name": "Aimee Dickerson",
      },
      Object {
        "id": 2,
        "name": "Tanisha Dickson",
      },
    ],
    "gender": "female",
    "greeting": "Hello, Robin Rodriguez! You have 6 unread messages.",
    "guid": "9b5c90e5-df6b-4195-bdb5-8ad4efa0ac03",
    "index": 11,
    "isActive": false,
    "latitude": 67.747371,
    "longitude": 139.090236,
    "name": "Robin Rodriguez",
    "phone": "+1 (899) 413-2671",
    "registered": "2016-08-14T03:43:48 +07:00",
    "tags": Array [
      "veniam",
      "ad",
      "sunt",
      "tempor",
      "consectetur",
      "minim",
      "nisi",
    ],
  },
]
`;

exports[`execute async handler should succeed with a returned value 1`] = `"foo"`;

exports[`execute callback handler should fail when returning an error 1`] = `[Error: Fail]`;

exports[`execute callback handler should succeed with a returned value 1`] = `"foo"`;

exports[`execute non-promise returned handler should resolve to null 1`] = `null`;
